source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1080,25 +1096,60 @@ enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
   const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
-
-  // Fallback on first certificate.
-  const TlsContext* selected_ctx = &tls_contexts_[0];
-  auto ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
-  for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable != ctx.is_ecdsa_) {
-      continue;
+  absl::string_view sni = absl::NullSafeStringView(
+      SSL_get_servername(ssl_client_hello->ssl, TLSEXT_NAMETYPE_host_name));
+
+  const TlsContext* selected_ctx = nullptr;
+
+  // do SNI matching and pkey type matching if SNI exists
+  if (!sni.empty()) {
+    PkeyTypesMap pkey_types_map;
+    // Match on exact server name, i.e. "www.example.com" for "www.example.com".
+    const auto server_name_exact_match = server_names_map_.find(sni);
+    if (server_name_exact_match != server_names_map_.end()) {
+      pkey_types_map = server_name_exact_match->second;
+    } else {
+      // Match on all wildcard domains, i.e. ".example.com" and ".com" for "www.example.com".
+      size_t pos = sni.find('.', 1);
+      while (pos < sni.size() - 1 && pos != std::string::npos) {
+        absl::string_view wildcard = sni.substr(pos);
+        const auto server_name_wildcard_match =
+            server_names_map_.find(static_cast<std::string>(wildcard));
+        if (server_name_wildcard_match != server_names_map_.end()) {
+          pkey_types_map = server_name_wildcard_match->second;
+          break;
+        }
+        pos = sni.find('.', pos + 1);
+      }
     }
 
-    auto action = ocspStapleAction(ctx, client_ocsp_capable);
-    if (action == OcspStapleAction::Fail) {
```

Comment:
I believe the current code loops through all the certificates (TlsContext instances) and picks the first one that is acceptable. These lines handle the case where a candidate TlsContext does not meet the OCSP requirements (e.g. the combination of the ocsp stapling policy and the certificates ocsp document which might be missing or expired). If the TlsContext doesn't meet the requirements, then the it continues searching and might select a later certificate in the same way that it continues if the certificate's public key type was not supported by the client.

Functionally this means:
1. if you have an ECDSA cert and an RSA cert configured, and
2. a client that supported both, and
3. the ECDSA cert fails the OCSP check, and
4. the RSA cert passes the OCSP check,
the current version will serve the RSA cert.

In the proposed version, I expect it would select the ECDSA cert and then return a cert error when the second OCSP check occurs.

Reply:
@zuercher

Thanks. This makes sense, I'd like add to proposed version for compatibility. But I think it's no need to use continue logic, since we only support at most one ecdsa and at most one rsa certificate.